src/main/java/com/visenze/productsearch/param/BaseProductSearchParam.java

```diff
@@ -0,0 +1,825 @@
+
+    /**
+     * Which SDK e.g. Javascript, Swift, Android, Java (for server side)
+     */
+    protected Optional<String> vaSdk = Optional.absent();
+
+    /**
+     * Store our tracking SDK’s version e.g. “1.0.2”. For debugging if there is
+     * a bug and also to measure SDK popularity/usage
+     */
+    protected Optional<String> vaSdkVersion = Optional.absent();
+
+    /**
+     * Custom mapping for future-proofing
+     */
+    protected Optional<Map<String, String>> customParams = Optional.absent();
+
+    /**
+     * Visenze Analytics fields
+     */
+    protected Optional<String> vaOs = Optional.absent();
+
+    /**
+     * Visenze Analytics fields
+     */
+    protected Optional<String> vaOsv = Optional.absent();
+
+    /**
+     * Visenze Analytics fields
+     */
+    protected Optional<String> vaDeviceBrand = Optional.absent();
+
+    /**
+     * Visenze Analytics fields
+     */
+    protected Optional<String> vaDeviceModel = Optional.absent();
+
+    /**
+     * Visenze Analytics fields
+     */
+    protected Optional<String> vaAppBundleId = Optional.absent();
+
+    /**
+     * Visenze Analytics fields
+     */
+    protected Optional<String> vaAppName = Optional.absent();
+
+    /**
+     * Visenze Analytics fields
+     */
+    protected Optional<String> vaAppVersion = Optional.absent();
+
+    /**
+     * Visenze Analytics fields
+     */
+    protected Optional<String> vaAaid = Optional.absent();
+
+    /**
+     * Visenze Analytics fields
+     */
+    protected Optional<String> vaDidmd5 = Optional.absent();
+
+    /**
+     * Visenze Analytics fields
+     */
+    protected Optional<Number> vaN1 = Optional.absent();
+
+    /**
+     * Visenze Analytics fields
+     */
+    protected Optional<Number> vaN2 = Optional.absent();
+
+    /**
+     * Visenze Analytics fields
+     */
+    protected Optional<String> vaS1 = Optional.absent();
+
+    /**
+     * Visenze Analytics fields
+     */
+    protected Optional<String> vaS2 = Optional.absent();
+
+    /**
+     * Convert this object into it's multimap representation.
+     *
+     * @return multimap of class variable to value
+     */
+    public Multimap<String, String> toMultimap() {
+        Multimap<String, String> multimap = HashMultimap.create();
+
+        if (page.isPresent())
+            multimap.put(PAGE, page.get().toString());
+
+        if (limit.isPresent())
+            multimap.put(LIMIT, limit.get().toString());
+
+        if (vaUid.isPresent())
+            multimap.put(VA_UID, vaUid.get());
+
+        if (vaSid.isPresent())
+            multimap.put(VA_SID, vaSid.get());
+
+        if (filters.isPresent()) {
+            for (Map.Entry<String, String> entry : filters.get().entrySet())
+                multimap.put(FILTERS, entry.getKey() + COLON + entry.getValue());
+        }
+
+        if (textFilters.isPresent()) {
+            for (Map.Entry<String, String> entry : textFilters.get().entrySet())
```

Comment:
from Codeclimate: identical code block between filters and textfilters , refactor into 1 function:

putMap(Multimap, String paramName, map)
e.g. putMap(multimap, FILTERS, filters) , putMap(multimap, TEXT_FILTERS, textFilters)